File: src/r_scripts/start_script.R
source("~/Desktop/Candidate Assessment/helpers.r")
source("~/Desktop/Candidate Assessment/convertMe.r")

install.packages("tidyverse")
install.packages("dplyr")
install.packages("readr")
install.packages("lubridate")

library("tidyverse")
library("dplyr")
library("readr")
library("lubridate")

dfCannibalizationFactors <- read.csv("~/Desktop/Candidate Assessment/data/dfCannibalizationFactors.csv")
dfMonthlySales <- read.csv("~/Desktop/Candidate Assessment/data/dfMonthlySales.csv")
dfReinvestmentFactors <- read.csv("~/Desktop/Candidate Assessment/data/dfReinvestmentFactors.csv")
dfReinvestmentProjects <- read.csv("~/Desktop/Candidate Assessment/data/dfReinvestmentProjects.csv")
dfSalesDaysFuture <- read.csv("~/Desktop/Candidate Assessment/data/dfSalesDaysFuture.csv")

# dfSalesDaysFuture cleaning
missing_date_index <- dfSalesDaysFuture$close_date == 'null'
dfSalesDaysFuture$close_date[missing_date_index] <- '2200-01-01'

dfSalesDaysFuture <- dfSalesDaysFuture %>% mutate(
  loc_num = str_pad(loc_num, 5, pad = "0"),
  open_date = ymd(open_date),
  close_date = ymd(close_date), 
  months_predict = 120
)

# dfCannibalizationFactors cleaning
dfCannibalizationFactors <- dfCannibalizationFactors %>% mutate(month = ymd(month))

# dfMonthlySales cleaning
dfMonthlySales <- dfMonthlySales %>% mutate(
  loc_num = str_pad(loc_num, 5, pad = "0"),
  month = ymd(month), 
  open_date = ymd(open_date)
)

# dfReinvestmentFactors cleaning
dfReinvestmentFactors <- dfReinvestmentFactors %>% mutate(
  loc_num = str_pad(loc_num, 5, pad = "0"),
  month = ymd(month))

# dfReinvestmentProjects cleaning
dfReinvestmentProjects <- dfReinvestmentProjects %>% mutate(
  loc_num = str_pad(loc_num, 5, pad = "0"),
  shutdown = ymd(shutdown), 
  reopen = ymd(reopen)
)

final_df <- convertMe(dfSalesDaysFuture, 
          dfMonthlySales, 
          dfReinvestmentProjects, 
          dfCannibalizationFactors, 
          dfReinvestmentFactors)

write.csv(final_df, "/Users/riddhi_gohil/Desktop/personal_git_repos/r-to-pyspark/output/r_final_sales_days_forecast.csv", row.names = FALSE)

File: src/python_scripts/start_script.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, lpad, to_date, lit, when
# from converted_pyspark import convert_me
from optmised_code import convert_me

# Initialize Spark session
spark = SparkSession. \
            builder \
            .appName("Start-Script") \
            .getOrCreate()

# TODO: Can put into different YAML file later on for extensibility
# Defining data paths
BASE_PATH = "data/"
data_files = {
    "dfCannibalizationFactors": "dfCannibalizationFactors.csv",
    "dfMonthlySales": "dfMonthlySales.csv",
    "dfReinvestmentFactors": "dfReinvestmentFactors.csv",
    "dfReinvestmentProjects": "dfReinvestmentProjects.csv",
    "dfSalesDaysFuture": "dfSalesDaysFuture.csv"
}

# Load all CSVs into a dictionary of DataFrames
dataFrames = {
    name: spark.read.option("header", True).option("inferSchema", True).csv(BASE_PATH + fname)
    for name, fname in data_files.items()
}
# print(dataFrames)

def format_dates_and_pad_loc(df, loc_col="loc_num", date_cols=None):
    """
    Pads key column named loc_col to 5 digits & 
    parses date columns.

    Args:
        df: Input DataFrame.
        loc_col: Column name to pad. Currently 'loc_num'.
        date_cols: List of date columns to convert.

    Returns:
        Transformed DataFrame.
    """
    if date_cols:
        for col_name in date_cols:
            df = df.withColumn(col_name, to_date(col(col_name)))
    return df.withColumn(loc_col, lpad(col(loc_col), 5, "0"))

# Clean and transform each DataFrame
dataFrames["dfSalesDaysFuture"] = (
        dataFrames["dfSalesDaysFuture"]
        .withColumn("close_date", when(col("close_date") == "null", "2200-01-01").otherwise(col("close_date")))
        .transform(lambda df: format_dates_and_pad_loc(df, date_cols=["open_date", "close_date"]))
        .withColumn("months_predict", lit(120))
    )

dataFrames["dfCannibalizationFactors"] = dataFrames["dfCannibalizationFactors"] \
                                            .withColumn("month", to_date("month"))

# TODO: Check if: 
#       close date will should be null
#       sales column values need to be rounded 
#       integers needed or float
dataFrames["dfMonthlySales"] = format_dates_and_pad_loc(
    dataFrames["dfMonthlySales"], date_cols=["month", "open_date"]
)

# TODO: Check if:
#      factor col should be float
dataFrames["dfReinvestmentFactors"] = format_dates_and_pad_loc(
    dataFrames["dfReinvestmentFactors"], date_cols=["month"]
)


dataFrames["dfReinvestmentProjects"] = format_dates_and_pad_loc(
    dataFrames["dfReinvestmentProjects"], date_cols=["shutdown", "reopen"]
)

# print("\nCleansed DataFrame: ")
# for name, df in dataFrames.items():
#     print(f"\n{name} schema:")
#     df.printSchema()
#     df.show(5, False) # comment later; cause lazy eval


final_df = convert_me(dataFrames["dfSalesDaysFuture"],
           dataFrames["dfMonthlySales"],
           dataFrames["dfReinvestmentProjects"],
           dataFrames["dfCannibalizationFactors"],
           dataFrames["dfReinvestmentFactors"])

final_df.coalesce(1) \
    .write \
    .mode("overwrite") \
    .option("header", True) \
    .csv("output/py_final_sales_days_forecast.csv")
